design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,190 @@
+    #  podAffinity:
+    #  podAntiAffinity:
+    # The requests and limits set here allow the ganesha pod(s) to use half
+    # of one CPU core and 1 gigabyte of memory
+    resources:
+    #  limits:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+    #  requests:
+    #    cpu: "500m"
+    #    memory: "1024Mi"
+
+  # Export settings
+  exports:
+    # The pseudoroot path. This is where the export will appear in the NFSv4
+    # pseudoroot namespace.
+  - pseudoPath: /cephfs/share1
+    # The directory in the exported file system this export is rooted on
+    path: /share1
+    allowedClients:
+    - clients:
+      - 172.17.0.0/16
+      - node1
+      accessMode: ReadOnly
+      squash: none
+
+  # NFS client recovery storage settings
+  clientRecovery:
+    # RADOS pool where NFS client recovery data is stored.
+    pool: myfs-data0
+    # RADOS namespace where NFS client recovery data is stored.
+    namespace: ganesha-ns
+
+```
+
+When the  nfs-ganesha.yaml is created the following will happen:
+
+- Rook's Ceph operator sees the creation of the NFS-Ganesha CRD and
+  stores the exports specified in the CRD as ConfigMaps.
+
+- The operator creates as many [k8s Deployments] as the number of active
+  Ganesha servers mentioned in the CRD. Each deployment brings up a Ganesha
+  server pod, a replicaset of size 1.
+
+- The ganesha servers, each running in a separate pod, use identical ganesha
+  config (ganesha.conf) that refer to exports stored as ConfigMaps.
+
+- The operator creates a k8s service for each of the ganesha server pods
+  to allow each of the them to have a stable IP address.
+
+The ganesha server pods constitute an active-active high availability NFS
+server cluster. If one of the active Ganesha server pods goes down, k8s brings
+up a replacment ganesha server pod with the same configuration and IP address.
+The NFS server cluster can be scaled up or down by updating the
+number of the active Ganesha servers in the CRD (using `kubectl edit` or
+modifying the original CRD and running `kubectly apply -f <CRD yaml file>`).
+
+### Dynamically adding, updating, and  removing exports of running Ganesha servers
+
+To dynamically update the exports of a Ganesha cluster you will have to update
+its CRD. On updating the CRD, the operator  will update the exports stored in
+the ConfigMaps, and will send SIGHUP signal or DBus message to the Ganesha
+servers to refresh their exports. Currently, there are issues with using either
+of the IPCs, SIGHUP and DBus, to dynamically update exports of the Ganesha
+servers running inside containers. But this should be fixed soon.
+
+### Consuming the NFS shares
+
+The external consumer will fetch the ganesha server IPs by querying the
+k8s services of the Ganesha server pods. It should have network access to the
+Ganesha pods to manually mount the shares using a NFS client. Later, support
+will be added to allow user pods to easily consume the NFS shares via PVCs,
+the steps to fetch the Ganesha
```

Comment:
This should be something the ceph-mgr can expose externally via the RESTful API or CLI. We don't want Manila or any other "end user" to need to know about Kubernetes/Rook.

Reply:
It should just look like a regular set of NFS servers. What is going to consume info from ceph-mgr here? Manila?

Ultimately, I think we eventually want to expose the whole topology of the NFS server cluster via NFSv4.1 fs_locations_info, but that will require some NFS client-side work to allow it to fetch and use that information. Even with that though, you'll need some way to get an initial address to mount.